sound/pci/hda/patch_hdmi.c

```diff
@@ -1703,30 +1767,36 @@ static void silent_stream_enable(struct hda_codec *codec,
 	/* configure unused pins to choose other converters */
 	pin_cvt_fixup(codec, per_pin, 0);
 
-	snd_hdac_sync_audio_rate(&codec->core, per_pin->pin_nid,
-				 per_pin->dev_id, I915_SILENT_RATE);
-
-	/* trigger silent stream generation in hw */
-	format = snd_hdac_calc_stream_format(I915_SILENT_RATE, I915_SILENT_CHANNELS,
-					     I915_SILENT_FORMAT, I915_SILENT_FORMAT_BITS, 0);
-	snd_hda_codec_setup_stream(codec, per_pin->cvt_nid,
-				   I915_SILENT_FMT_MASK, I915_SILENT_FMT_MASK, format);
-	usleep_range(100, 200);
-	snd_hda_codec_setup_stream(codec, per_pin->cvt_nid, I915_SILENT_FMT_MASK, 0, format);
-
-	per_pin->channels = I915_SILENT_CHANNELS;
-	hdmi_setup_audio_infoframe(codec, per_pin, per_pin->non_pcm);
+	if (spec->silent_stream_type == SILENT_STREAM_I915)
+		silent_stream_enable_i915(codec, per_pin);
+	else
+		silent_stream_enable_kae(codec, per_pin);
 
  unlock_out:
 	mutex_unlock(&per_pin->lock);
+
+	/* in i915 silent-stream mode, codec must be kept in D0 */
+	if (err || spec->silent_stream_type != SILENT_STREAM_I915)
```

Comment:
same here, if the type if OFF do you need to do anything?

Reply:
Ditto here, I tried to make this clearer.